state-machine/src/commonMain/kotlin/com/freeletics/mad/statemachine/StateMachine.kt

```diff
@@ -0,0 +1,23 @@
+package com.freeletics.mad.statemachine
+
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * A state machine that emits [State] objects through the [Flow] returned by [state]. The state
+ * can be mutated through actions passed to [dispatch].
+ */
+interface StateMachine<State : Any, Action : Any> {
+
+    /**
+     * A [Flow] that emits the current state as well as all changes to the state.
+     */
+    val state: Flow<State>
+
+    /**
+     * An an [Action] to the [StateMachine]. The implementation can mutate the [State] based on
+     * these actions or trigger side effects.
+     */
+    suspend fun dispatch(action: Action)
```

Comment:
For renderers I would see that happen in the glue code like you mentioned. The `Renderer` just exposes a `Flow` so when collecting it `dispatch` will be called in a suspendable context. The `sendAction` convenience method inside `ViewRenderer` is not suspending and the internal `MutableSharedFlow` would buffer actions if `dispatch` suspends while new actions are emitted.

For compose I've been doing something like this. The actual composable containing the ui for the screen looks like this:
```
@Composable
fun FooScreen(
    state: FooState,
    onAction: @Composable (FooAction) -> Unit
) { ... }
```
and then have a wrapper composable that contains the glue and launches the dispatch
```
val state by stateMachine.state.collectAsState()
FooScreen(state) {
    LaunchedEffect(stateMachine) { stateMachine.dispatch(it) }
}
```

The compose snippet makes me wonder whether `state` should be a `StateFlow` though. There is `StateFlow<T>.collectAsState()` and `Flow<T>.collectAsState(initial: T). So if we only expose a `Flow` from the `StateMachine` we'd need to provide an initial value at the collection point which imo is the wrong place.

Reply:
For Compose and FlowRedux (but could be for `StateMachine` interface, I was thinking about something like this:

```kotlin
val (state : MutableState<State>, dispatcher : (Action) -> Unit) = rememberStateMachine(MyStateMachine)

FooScreen(state, dispatcher)
```

So the idea is that `rememberStateMachine()` would take care of coroutineScope management / subscription management so that you as developer don't need to take care of managing coroutineScope (so Flow) is entirely hidden nor to launch from a coroutine scope to dispatch actions but rather you have a non suspending dispatcher that you can invoke anytime you want and pass down through the compose hierarchy tree.

Reply:
Even then you'd need to either have a `StateFlow` or pass the initial state to `rememberStateMachine` otherwise you can't create a compose `State` from what I can see.